### PR TITLE
8341091: CDS: Segmented roots array misses roots

### PR DESCRIPTION
**Pull request by shipilev**
There are a few bugs in [JDK-8338912](https://bugs.openjdk.org/browse/JDK-8338912) that we discovered in Leyden that uses multiple segments: a) root_index in ArchiveHeapWriter::copy_roots_to_buffer is initialized to zero for every segment, so all segments have the same set of roots; b) HeapShared::get_root and friends compute segment and internal indices modulo _byte_ size, while they should instead do it modulo _element_ size.

This does not affect current mainline directly, since we run with a single segment the overwhelming majority of the time.

Additional testing:
 - [x] Linux x86_64 server fastdebug, `runtime/cds`
 - [x] Linux x86_64 server fastdebug, `runtime/cds` + `-XX:-UseCompressedOops`
 - [x] Transplanted the patch to Leyden repo and verified it passes `runtime/cds`
 - [x]  Linux x86_64 server fastdebug, `all`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341091](https://bugs.openjdk.org/browse/JDK-8341091): CDS: Segmented roots array misses roots (**Bug** - P3)

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) Review applies to [c0635da3](https://git.openjdk.org/jdk/pull/21221/files/c0635da3dc0403a60583d8517f2581806e16cb24)
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21221/head:pull/21221` \
`$ git checkout pull/21221`

Update a local copy of the PR: \
`$ git checkout pull/21221` \
`$ git pull https://git.openjdk.org/jdk.git pull/21221/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21221`

View PR using the GUI difftool: \
`$ git pr show -t 21221`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21221.diff">https://git.openjdk.org/jdk/pull/21221.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21221#issuecomment-2379575617)

### CODE REVIEW

**Comment by shipilev** [comment]
@iklam ^

**Review by iklam — APPROVED** [review]
LGTM

**Comment by shipilev** [comment]
Thank you! This is Hotspot patch, so I need another reviewer for this one.

File: src/hotspot/share/cds/archiveHeapLoader.cpp [hunk 1]
```diff
@@ -378,11 +378,10 @@ void ArchiveHeapLoader::finish_initialization() {
     // The heap roots are stored in one or more segments that are laid out consecutively.
     // The byte size of each segment (except for the last one) is max_size.
```
Comment (adinn) [inline src/hotspot/share/cds/archiveHeapLoader.cpp:379]: Comment here mentions `max_size` but that doesn't exist any more.

**Review by adinn — APPROVED** [review]
Looks ok, apart from one minor tweak to a comment

File: src/hotspot/share/cds/archiveHeapLoader.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/share/cds/archiveHeapLoader.cpp]: Thanks! I tightened up the comment a bit.

**Review by adinn — APPROVED** [review]
Good!

**Comment by shipilev** [comment]
Thanks! 

/integrate

**Pull request closed without merging**